### DISCUSSION TITLE
Tests failed on receive a response -The "url" argument must be of type string. Received undefined

### DISCUSSION THREAD

**Discussion by martynaol** [pactumjs/pactum#131] [Q&A] [answered]
I have : 
```
const { Given, When, Then, Before } = require('@cucumber/cucumber');
const { spec, request } = require('pactum');

Given('I have the layout', async () => {
  await spec()
    .get('/layouts')
    .expectStatus(200)
    .stores('layoutId', 'data[0].id');
});

When('I assign the layout to the ad unit', async () => {
  await spec().post(setLayoutUrl).withBody({
    layoutId: '$S{layoutId}',
  });

When('I receive a response', async () => {
  await spec().toss();
});

Then(
  'I have successfully assign to ad unit, the response should have a status {int}',
  code => {
    spec().response().should.have.status(code);
  }
);
});

```
**Test failed**
` Scenario: Assign a layout to an ad unit # test/api/features/assign-layout.feature:4
   ✔ Given I have the layout # test/api/features/step_definitions/assignLayout.js:51
   ✔ When I assign the layout to the ad unit # test/api/features/step_definitions/assignLayout.js:60
   ✖ And I receive a response # test/api/features/step_definitions/assignLayout.js:83
       TypeError [ERR_INVALID_ARG_TYPE]: The "url" argument must be of type string. Received undefined
           at validateString (internal/validators.js:120:11)
           at Url.parse (url.js:160:3)
           at Object.urlParse [as parse] (url.js:155:13)
           at setBaseUrl (/home/martyna-olesinska/projects/postclick/packages/automated-layouts-service/node_modules/pactum/src/helpers/requestProcessor.js:30:20)
           at Object.process (/home/martyna-olesinska/projects/postclick/packages/automated-layouts-service/node_modules/pactum/src/helpers/requestProcessor.js:12:5)
           at Tosser.toss (/home/martyna-olesinska/projects/postclick/packages/automated-layouts-service/node_modules/pactum/src/models/Tosser.js:29:39)
           at Spec.toss (/home/martyna-olesinska/projects/postclick/packages/automated-layouts-service/node_modules/pactum/src/models/Spec.js:477:19)
           at World.<anonymous> (/home/martyna-olesinska/projects/postclick/packages/automated-layouts-service/test/api/features/step_definitions/assignLayout.js:84:16)
   - Then I have successfully assign to ad unit, the response should have a status 200 # test/api/features/step_definitions/assignLayout.js:87
   - And I have successfully assigned layout and content to the ad unit # test/api/features/step_definitions/assignLayout.js:94
`

Currently I use mix of approaches. So I have : 
const pactum = require('pactum');
let spec = pactum.spec() 
Sometimes inside the test I use
```
`await pactum
    .spec()
    .get(`/clients/)
    .withQueryParams({
      level: 'adgroup',
    })
    .expectStatus(200)
    .returns('data[0].id');`
```
or 
```
spec = pactum.spec();
  await spec.post('/set-layout).withBody({
    layoutId: null,
    pageId: null,
  });
```
Its works, but I'm curious what is a difference. Especially, I need re-initial spec each time

**Comment by ASaiAnudeep** [comment, marked as answer]
The method `spec` will create an instance through which we can make a single request. Please refer [this](https://pactumjs.github.io/reference/spec) for more usage details.

```js
const spec_one = spec();
const spec_two = spec();
```

The method `toss` will execute the spec and returns a promise. Please refer [this](https://pactumjs.github.io/reference/toss) for more usage details.

```js
await spec().get('/api/users').expectStatus(200).toss();
await spec().get('/api/projects/p1').expectStatus(404).toss();
```

In your step definitions file, you have created an empty spec and running without defining any request details.

```js
When('I receive a response', async () => {
  await spec().toss();  // empty spec, this will throw an error
});
```

**Reply by martynaol** [reply]
Thank you very much for your fast reply. Now, things seems clearly
